test/e2e/service.go

```diff
@@ -836,6 +836,167 @@ var _ = ginkgo.Describe("Services", func() {
+				serviceName,
+				v1.IPFamilyPolicyPreferDualStack,
+				endpointHTTPPort,
+				endpointUDPPort,
+				clusterHTTPPort,
+				clusterUDPPort,
+				endpointsSelector,
+				v1.ServiceExternalTrafficPolicyTypeCluster,
+			)
+			service, err = f.ClientSet.CoreV1().Services(f.Namespace.Name).Create(context.Background(), service, metav1.CreateOptions{})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Waiting for the endpoints to pop up")
+			err = framework.WaitForServiceEndpointsNum(
+				context.TODO(),
+				f.ClientSet,
+				f.Namespace.Name,
+				serviceName,
+				1,
+				time.Second,
+				wait.ForeverTestTimeout,
+			)
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Creating an external client")
+			clientIPv4, clientIPv6 := createClusterExternalContainer(
+				clientContainerName,
+				agnhostImage,
+				[]string{"--privileged", "--network", "kind"},
+				[]string{"pause"},
+			)
+
+			clientIP := clientIPv4
+			serverNodeIP := serverNodeIPv4
+			ipContainerCmd := "ip"
+			if IsIPv6Cluster(f.ClientSet) {
+				clientIP = clientIPv6
+				serverNodeIP = serverNodeIPv6
+				ipContainerCmd = "ip -6"
+			}
+
+			const pmtu = "1300"
+			payloads := map[string]string{
+				"non-fragmented": "1220",
+				"fragmented":     "1320",
+			}
+
+			// We set a route MTU towards the server node emulating that PMTUD
+			// has already happened resulting in a plausible low PMTU.
+			// For UDP the system wide default IP_PMTUDISC_WANT will
+			// result in fragmentation for packets bigger than the PMTU.
+			// Note that fragmentation could also happen if a client chooses to
+			// not use PMTUD with IP_PMTUDISC_DONT both for TCP or UDP but the
+			// test setup required to achieve fragmentation without emulating
+			// PMTUD is more complex so we stick to UDP.
+			ginkgo.By("Lowering PMTU towards the server")
+			ipContainerCmd += " route add " + serverNodeIP + " dev eth0 src " + clientIP + " mtu " + pmtu
+			cmd := []string{
+				containerRuntime,
+				"exec",
+				clientContainerName,
+				"/bin/sh",
+				"-c",
+				ipContainerCmd,
+			}
+			framework.Logf("Running %v", cmd)
+			_, err = runCommand(cmd...)
+			framework.ExpectNoError(err, "lowering MTU in the external kind container failed: %v", err)
+
+			var udpPort int32
+			for _, port := range service.Spec.Ports {
+				if port.Protocol == v1.ProtocolUDP {
+					udpPort = port.NodePort
+				}
+			}
+			gomega.Expect(udpPort).NotTo(gomega.Equal(0))
+
+			// To check that forwarding did not happen via host, we send a
+			// non-fragmented packet first and then a fragmented one on the same
+			// source port. If the server sees the same source port it means
+			// that both packets were forwarded the same. This is because when
+			// forwarding via OVN, DNAT happens directly to the endpoint IP,
+			// whereas forwarding via host, DNAT happens to the service cluster
+			// IP and then forwaded to OVN. Thus, OVN sees both as different
+			// streams and ends up replacing the source port to avoid the
```

Comment:
> Is zone 16 where the SNAT will choose a different non-overlapping port?

Yeah, this is a real example:

ovn path
```
udp      17 23 src=172.18.0.5 dst=172.18.0.4 sport=45083 dport=30648 src=10.244.1.4 dst=172.18.0.5 sport=90 dport=45083 mark=10 secctx=system_u:object_r:unlabeled_t:s0 zone=2 use=1
udp      17 23 src=172.18.0.5 dst=10.244.1.4 sport=45083 dport=90 src=10.244.1.4 dst=100.64.0.2 sport=90 dport=45083 mark=0 secctx=system_u:object_r:unlabeled_t:s0 use=1
udp      17 23 src=100.64.0.2 dst=10.244.1.4 sport=45083 dport=90 src=10.244.1.4 dst=100.64.0.2 sport=90 dport=45083 mark=0 secctx=system_u:object_r:unlabeled_t:s0 zone=20 use=1
```
host path
```
udp      17 25 src=172.18.0.5 dst=172.18.0.4 sport=45083 dport=30648 src=10.96.150.92 dst=172.18.0.5 sport=91 dport=45083 mark=0 secctx=system_u:object_r:unlabeled_t:s0 use=1
udp      17 25 src=172.18.0.5 dst=10.96.150.92 sport=45083 dport=91 src=10.96.150.92 dst=169.254.169.2 sport=91 dport=45083 mark=0 secctx=system_u:object_r:unlabeled_t:s0 zone=64001 use=1
udp      17 25 src=169.254.169.2 dst=10.96.150.92 sport=45083 dport=91 src=10.244.1.4 dst=169.254.169.2 sport=90 dport=45083 mark=10 secctx=system_u:object_r:unlabeled_t:s0 zone=2 use=1
udp      17 25 src=169.254.169.2 dst=10.244.1.4 sport=45083 dport=90 src=10.244.1.4 dst=100.64.0.2 sport=90 dport=14591 mark=0 secctx=system_u:object_r:unlabeled_t:s0 use=1
```

I guess it could be more clear if I attribute it to the masquerading instead. I will try that but let me know what would be more clear here if you have an idea.

> Also, how is this applicable, because without your fix, the first fragmented packet would go via OVN. The second fragmented packet would go by host, but it has no L4 header, so isnt it not even DNAT'ed by OVN?

It's not like that. Non-fragmented packets go through OVN, fragmented packets go through host. So the host reassembles the fragments and does have L4 header. So the issue is when there is a mix of fragmented and non fragmented packets in a connection stream.

Reply:
Note that OVS in the normal fragmentation mode, the first fragment does not have L4 headers either:

```
       set-frags switch frag_mode
              Configures switch's treatment of IPv4 and IPv6 fragments.
              The choices for frag_mode are:

              normal Fragments pass through the flow table like non-
                     fragmented packets.  The TCP ports, UDP ports, and
                     ICMP type and code fields are always set to 0, even
                     for fragments where that information would
                     otherwise be available (fragments with offset 0).
                     This is the default fragment handling mode for an
                     OpenFlow switch.

              drop   Fragments are dropped without passing through the
                     flow table.

              reassemble
                     The switch reassembles fragments into full IP
                     packets before passing them through the flow table.
                     Open vSwitch does not implement this fragment
                     handling mode.

              nx-match
                     Fragments pass through the flow table like non-
                     fragmented packets.  The TCP ports, UDP ports, and
                     ICMP type and code fields are available for
                     matching for fragments with offset 0, and set to 0
                     in fragments with nonzero offset.  This mode is a
                     Nicira extension.
```

Reply:
Thanks @jcaamano this clarifies it.